### PR TITLE
Xds v3

### PR DESCRIPTION
**Pull request by EItanya**
# Description

Update Gloo to use envoy V3 APIs across the board

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/cli/pkg/xdsinspection/get_last_config.go [hunk 1]
```diff
@@ -294,9 +298,11 @@ func (xd *XdsDump) String() string {
...
 func toYaml(pb proto.Message) ([]byte, error) {
-	jsn, err := protoutils.MarshalBytes(pb)
+	buf := &bytes.Buffer{}
+	jpb := &jsonpb.Marshaler{}
+	err := jpb.Marshal(buf, pb)
 	if err != nil {
 		return nil, err
 	}
-	return yaml.JSONToYAML(jsn)
+	return yaml.JSONToYAML(buf.Bytes())
```
Comment (yuval-k) [inline projects/gloo/cli/pkg/xdsinspection/get_last_config.go:307]: why this change?

File: projects/gloo/pkg/plugins/hcm/plugin.go [hunk 2]
```diff
@@ -44,7 +44,7 @@ func (p *Plugin) RegisterHcmPlugins(allPlugins []plugins.Plugin) {
 // ProcessListener has two responsibilities:
 // 1. apply the core HCM settings from the HCM plugin to the listener
 // 2. call each of the HCM plugins to make sure that they have a chance to apply their modifications to the listener
-func (p *Plugin) ProcessListener(params plugins.Params, in *v1.Listener, out *envoyapi.Listener) error {
+func (p *Plugin) ProcessListener(params plugins.Params, in *v1.Listener, out *envoy_config_listener_v3.Listener) error {
```
Comment (yuval-k) [inline projects/gloo/pkg/plugins/hcm/plugin.go:47]: this plugin might need some more work WRT version of hcm

**Review by yuval-k — COMMENTED** [review]
reviewed about half; leaving some comments and will resume after lunch

File: projects/gloo/cli/pkg/xdsinspection/get_last_config.go (hunk 1 above)
Comment (EItanya) [reply projects/gloo/cli/pkg/xdsinspection/get_last_config.go:307]: Was using gogo before, we need to use golang here

File: pkg/utils/gogoutils/health_check.go [hunk 3]
```diff
@@ -132,48 +132,35 @@ func ToEnvoyHealthCheck(check *envoycore_gloo.HealthCheck, secrets *v1.SecretLis
...
-				UseHttp2:               typed.HttpHealthCheck.GetUseHttp2(),
-				ServiceName:            typed.HttpHealthCheck.GetServiceName(),
-				RequestHeadersToAdd:    requestHeadersToAdd,
-				RequestHeadersToRemove: typed.HttpHealthCheck.GetRequestHeadersToRemove(),
-				ExpectedStatuses:       ToEnvoyInt64RangeList(typed.HttpHealthCheck.GetExpectedStatuses()),
+		hc.HealthChecker = &envoy_config_core_v3.HealthCheck_HttpHealthCheck_{
+			HttpHealthCheck: &envoy_config_core_v3.HealthCheck_HttpHealthCheck{
+				Host:                             typed.HttpHealthCheck.GetHost(),
+				Path:                             typed.HttpHealthCheck.GetPath(),
+				HiddenEnvoyDeprecatedUseHttp2:    typed.HttpHealthCheck.GetUseHttp2(),
```
[the remark this answers is not in the document]
Comment (EItanya) [reply pkg/utils/gogoutils/health_check.go:139]: This is the direct translation, I'll look into what this field has become in the newer version

File: projects/gloo/pkg/xds/internal/listener.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+	if filter == nil {
+		return nil
+	}
+	return &envoy_api_v2_listener.ListenerFilter{
+		Name: filter.GetName(),
+		ConfigType: &envoy_api_v2_listener.ListenerFilter_TypedConfig{
+			TypedConfig: filter.GetTypedConfig(),
+		},
+		// Skipping for now as we don't expose this field in our API, and it's recursion makes it more complex
+		FilterDisabled: nil,
```
Comment (yuval-k) [inline projects/gloo/pkg/xds/internal/listener.go:223]: good idea to skip

File: projects/gloo/pkg/utils/paths.go [hunk 5]
```diff
@@ -17,14 +17,14 @@ func PathAsString(matcher *matchers.Matcher) string {
...
-func EnvoyPathAsString(matcher *route.RouteMatch) string {
+func EnvoyPathAsString(matcher *envoy_config_route_v3.RouteMatch) string {
 	switch path := matcher.GetPathSpecifier().(type) {
-	case *route.RouteMatch_Prefix:
+	case *envoy_config_route_v3.RouteMatch_Prefix:
 		return path.Prefix
-	case *route.RouteMatch_Path:
+	case *envoy_config_route_v3.RouteMatch_Path:
 		return path.Path
-	case *route.RouteMatch_Regex:
```
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gloo/pkg/utils/paths.go:26]: It's used only in the GRPC plugin. This field isn't present in the V3 api, but during our translation to V3 route we upgrade normal regex to safe regex. So this should be fine.

File: projects/gloo/pkg/translator/defaults.go [hunk 6]
```diff
@@ -4,12 +4,6 @@ import (
 	"time"
 )
 
-const (
-	SslCertificateChainKey = "tls.crt"
-	SslPrivateKeyKey       = "tls.key"
-	SslRootCaKey           = "tls.root"
-)
```
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gloo/pkg/translator/defaults.go:11]: Yes, we are using the kubernetes tls secret constants now

File: projects/gloo/pkg/xds/envoy_snapshot.go [hunk 7]
```diff
@@ -100,14 +166,18 @@ func (s *EnvoySnapshot) GetResources(typ string) cache.Resources {
...
-	case RouteType:
+	case resource.RouteTypeV3:
 		return s.Routes
-	case ListenerType:
+	case resource.ListenerTypeV3:
 		return s.Listeners
+	case resource.ClusterTypeV2:
+		return s.hiddenDeprecatedClusters
+	case resource.ListenerTypeV2:
+		return s.hiddenDeprecatedListeners
```
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gloo/pkg/xds/envoy_snapshot.go:180]: Yes we do as it turns out, for rest xds v2

File: projects/gloo/pkg/xds/internal/endpoint.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+		return nil
+	}
+
+	downgraded := &envoy_api_v2.ClusterLoadAssignment_Policy{
+		DropOverloads: make(
+			[]*envoy_api_v2.ClusterLoadAssignment_Policy_DropOverload, 0, len(policy.GetDropOverloads()),
+		),
+		OverprovisioningFactor:  policy.GetOverprovisioningFactor(),
+		EndpointStaleAfter:      policy.GetEndpointStaleAfter(),
+		DisableOverprovisioning: policy.GetHiddenEnvoyDeprecatedDisableOverprovisioning(),
```
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gloo/pkg/xds/internal/endpoint.go]: Good point, those fields are unused now

**Comment by EItanya** [comment]
NOTE: Need to keep all AccessLog related APIs at envoy API V2 temporarily. The reason being that the version of go-control-plane we depend on does not have the `transport_api_version` field yet. We cannot bump this version until we get rid of gogo.

**Review by yuval-k — COMMENTED** [review]
looks good

**Review by yuval-k — APPROVED** [review]
